### PR TITLE
PR for #1394 custom error type for client/da-availability

### PR DESCRIPTION
**Pull request by peyha**
Here is my first draft for the custom error handling in the client/data-availability crate. (#1394)
I propose to use a DaError struct which is composed of a DaLayer and a custom DaLayerError type which encompasses several types of error (blob related, client related or config related).
I implemented it for the avail DA as an example (and will do it for Celestia and Ethereum after some feedback)

Please add the labels corresponding to the type of changes your PR introduces:

- Feature

## Does this introduce a breaking change?

I don't think so

### CODE REVIEW

**Review by EvolveArt — CHANGES_REQUESTED** [review]
Thanks for this PR. The goal is to remove completely string and mostly rely on enums with clearer names such as 'DaError::Config::FailedOpening(String)'

**Comment by peyha** [comment]
Ok, so if I understand correctly, I can remove the DALayer type attribute and then use even more customized error types ?

File: crates/client/data-availability/src/lib.rs [hunk 1]
```diff
@@ -43,6 +44,41 @@ pub enum DaLayer {
...
+    #[error("failed submitting data through client: {0}")]
+    FailedSubmission(String),
+    #[error("failed fetching data through client: {0}")]
+    FailedFetching(String),
+}
+
+#[derive(Error, Debug)]
+pub enum ConfigErr{
+    #[error("failed opening config: {0}")]
+    FailedOpening(String),
```
Comment (tdelabro) [inline crates/client/data-availability/src/lib.rs:62]: here replace String by std::fs:::Error

File: crates/client/data-availability/src/lib.rs [hunk 2]
```diff
@@ -43,6 +44,41 @@ pub enum DaLayer {
...
+    #[error("failed parsing config: {0}")]
+    FailedParsing(String),
```
Comment (tdelabro) [inline crates/client/data-availability/src/lib.rs:64]: And here by `serde::Errro` or `ParseError`, whatever the original type was

File: crates/client/data-availability/src/lib.rs [hunk 3]
```diff
@@ -43,6 +44,41 @@ pub enum DaLayer {
...
+#[derive(Error, Debug)]
+pub enum ClientErr{
+    #[error("failed connecting http: {0}")]
+    FailedHttpConnection(String),
+    #[error("failed connecting websocket: {0}")]
+    FailedWsConnection(String),
+    #[error("failed submitting data through client: {0}")]
+    FailedSubmission(String),
+    #[error("failed fetching data through client: {0}")]
+    FailedFetching(String),
```
Comment (tdelabro) [inline crates/client/data-availability/src/lib.rs:56]: Same here. Keep the original errors type

**Review by tdelabro — CHANGES_REQUESTED** [review]

**Comment by peyha** [comment]
It's a bit complicated to use the original error types, for instance FailedWsConnection can be linked with an anyhow::Error and a subxt::error::Error, should I use a more general Error type ?

**Comment by tdelabro** [comment]
` anyhow::Error ` is already a wrapper around any error type
You can use it to avoid creating new intermediate enum error

**Comment by peyha** [comment]
I factorized the error types as much as possible and re-used a known error type when possible

File: crates/client/data-availability/src/ethereum/mod.rs [hunk 4]
```diff
@@ -87,20 +86,20 @@ impl TryFrom<config::EthereumConfig> for EthereumClient {
 
     fn try_from(conf: config::EthereumConfig) -> Result<Self, Self::Error> {
         if !is_valid_http_endpoint(&conf.http_provider) {
-            return Err(format!("invalid http endpoint, received {}", &conf.http_provider));
+            return Err(DaError::FailedBuildingClient(anyhow::Error::msg("invalid http endpoint")).to_string());
```
Comment (tdelabro) [inline crates/client/data-availability/src/ethereum/mod.rs:89]: just une `anyhow::anyhow!("My string {}", my_var)`

File: crates/client/data-availability/src/ethereum/mod.rs [hunk 5]
```diff
@@ -87,20 +86,20 @@ impl TryFrom<config::EthereumConfig> for EthereumClient {
...
         }
 
-        let provider = Provider::<Http>::try_from(conf.http_provider).map_err(|e| format!("ethereum error: {e}"))?;
+        let provider = Provider::<Http>::try_from(conf.http_provider).map_err(|e| DaError::FailedBuildingClient(e.into()).to_string())?;
 
         let wallet: LocalWallet = conf
             .sequencer_key
             .parse::<LocalWallet>()
-            .map_err(|e| format!("ethereum error: {e}"))?
+            .map_err(|e| DaError::FailedConversion(e.into()).to_string())?
```
Comment (tdelabro) [inline crates/client/data-availability/src/ethereum/mod.rs:97]: You are once again storing a string here. Get rid of the `to_string`

File: crates/client/data-availability/src/ethereum/mod.rs [hunk 6]
```diff
@@ -87,20 +86,20 @@ impl TryFrom<config::EthereumConfig> for EthereumClient {
...
             .with_chain_id(conf.chain_id);
 
         let signer = Arc::new(SignerMiddleware::new(provider.clone(), wallet));
 
-        let cc_address: Address = conf.core_contracts.parse().map_err(|e| format!("ethereum error: {e}"))?;
+        let cc_address: Address = conf.core_contracts.parse::<Address>().map_err(|e| DaError::FailedConversion(e.into()).to_string())?;
```
Comment (tdelabro) [inline crates/client/data-availability/src/ethereum/mod.rs:102]: get rind of `to_string`

File: crates/client/data-availability/src/ethereum/mod.rs [hunk 7]
```diff
@@ -87,20 +86,20 @@ impl TryFrom<config::EthereumConfig> for EthereumClient {
 
     fn try_from(conf: config::EthereumConfig) -> Result<Self, Self::Error> {
         if !is_valid_http_endpoint(&conf.http_provider) {
-            return Err(format!("invalid http endpoint, received {}", &conf.http_provider));
+            return Err(DaError::FailedBuildingClient(anyhow::Error::msg("invalid http endpoint")).to_string());
         }
 
-        let provider = Provider::<Http>::try_from(conf.http_provider).map_err(|e| format!("ethereum error: {e}"))?;
+        let provider = Provider::<Http>::try_from(conf.http_provider).map_err(|e| DaError::FailedBuildingClient(e.into()).to_string())?;
```
Comment (tdelabro) [inline crates/client/data-availability/src/ethereum/mod.rs:92]: get rid of to_string

File: crates/client/data-availability/src/ethereum/config.rs [hunk 8]
```diff
@@ -33,8 +33,12 @@ impl TryFrom<&PathBuf> for EthereumConfig {
...
 
     fn try_from(path: &PathBuf) -> Result<Self, Self::Error> {
-        let file = File::open(path).map_err(|e| format!("error opening da config: {e}"))?;
-        serde_json::from_reader(file).map_err(|e| format!("error parsing da config: {e}"))
+        let file = File::open(path).map_err(|e|
+            DaError::FailedOpeningConfig(e).to_string()
+        )?;
+        serde_json::from_reader(file).map_err(|e| 
+            DaError::FailedParsingConfig(e).to_string()
+        )
```
Comment (tdelabro) [inline crates/client/data-availability/src/ethereum/config.rs:41]: get rid of `to_string`

**Review by tdelabro — CHANGES_REQUESTED** [review]

**Comment by peyha** [comment]
I just removed the String (it created an error in `crates/node/src/service.rs` which I adapted accordingly even though this part of the code has been deleted since then)

File: crates/node/src/service.rs [hunk 9]
```diff
@@ -442,8 +442,8 @@ pub fn new_full(
                 Arc::new(CelestiaClient::try_from(celestia_conf).map_err(|e| ServiceError::Other(e.to_string()))?)
             }
             DaLayer::Ethereum => {
-                let ethereum_conf = EthereumConfig::try_from(&da_path)?;
-                Arc::new(EthereumClient::try_from(ethereum_conf)?)
+                let ethereum_conf = EthereumConfig::try_from(&da_path).map_err(|e| e.to_string())?;
```
Comment (tdelabro) [inline crates/node/src/service.rs:445]: Sill a few `to_string` there

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/node/src/service.rs (hunk 9 above)
Comment (peyha) [reply crates/node/src/service.rs:445]: yes but the `new_full` function returns a `Result<TaskManager, ServiceError>` type so the `DaError` has to be converted (unless the ServiceError type includes DaError type)

Comment (tdelabro) [reply crates/node/src/service.rs:445]: Oh sure sorry

File: crates/client/data-availability/src/lib.rs [hunk 10]
```diff
@@ -43,6 +44,24 @@ pub enum DaLayer {
...
+    #[error("failed converting parameter: {0}")]
+    FailedConversion(anyhow::Error),
+    #[error("failed building client: {0}")]
+    FailedBuildingClient(anyhow::Error),
+    #[error("failed submitting data through client: {0}")]
+    FailedDataSubmission(anyhow::Error),
+    #[error("failed fetching data through client: {0}")]
+    FailedDataFetching(anyhow::Error),
+    #[error("failed validating data: {0}")]
+    FailedDataValidation(anyhow::Error),
```
Comment (tdelabro) [inline crates/client/data-availability/src/lib.rs:62]: Add an `InvalidHttpEndpoint` variant here

File: crates/client/data-availability/src/ethereum/mod.rs [hunk 11]
```diff
@@ -83,24 +82,24 @@ impl DaClient for EthereumClient {
 }
 
 impl TryFrom<config::EthereumConfig> for EthereumClient {
-    type Error = String;
+    type Error = DaError;
 
     fn try_from(conf: config::EthereumConfig) -> Result<Self, Self::Error> {
         if !is_valid_http_endpoint(&conf.http_provider) {
-            return Err(format!("invalid http endpoint, received {}", &conf.http_provider));
+            return Err(DaError::FailedBuildingClient(anyhow::anyhow!("invalid http endpoint, received {}", &conf.http_provider)));
```
Comment (tdelabro) [inline crates/client/data-availability/src/ethereum/mod.rs:89]: So you can use it here

**Review by tdelabro — CHANGES_REQUESTED** [review]

**Comment by peyha** [comment]
should be good now

File: crates/client/data-availability/src/lib.rs [hunk 12]
```diff
@@ -60,6 +60,8 @@ pub enum DaError {
     FailedDataFetching(anyhow::Error),
     #[error("failed validating data: {0}")]
     FailedDataValidation(anyhow::Error),
+    #[error("Invalid http endpoint: {0}")]
+    InvalidHttpEndpoint(String)
```
Comment (tdelabro) [inline crates/client/data-availability/src/lib.rs:64]: I think http_provider is of type Url

**Review by tdelabro — APPROVED** [review]

**Comment by tdelabro** [comment]
@peyha still some conflicts

**Comment by tdelabro** [comment]
@peyha you should use rebase rather than merge in this type of situation

**Comment by peyha** [comment]
wanted to use rebase to keep the branch clean but merged by error, shall i change something ?

File: crates/node/src/commands/run.rs [hunk 13]
```diff
@@ -163,8 +163,8 @@ fn init_da_client(da_layer: DaLayer, da_path: PathBuf) -> Result<Box<dyn DaClien
             Box::new(CelestiaClient::try_from(celestia_conf).map_err(|e| sc_cli::Error::Input(e.to_string()))?)
         }
         DaLayer::Ethereum => {
-            let ethereum_conf = EthereumConfig::try_from(&da_path)?;
-            Box::new(EthereumClient::try_from(ethereum_conf)?)
+            let ethereum_conf = EthereumConfig::try_from(&da_path).map_err(|e| e.to_string())?;
+            Box::new(EthereumClient::try_from(ethereum_conf).map_err(|e| e.to_string())?)
```
Comment (tdelabro) [inline crates/node/src/commands/run.rs:167]: Use `.map_err(|e| sc_cli::Error::Other(e.to_string()))?)` it's more explicit

**Review by tdelabro — CHANGES_REQUESTED** [review]

**Comment by tdelabro** [comment]
> wanted to use rebase to keep the branch clean but merged by error, shall i change something ?

No, we will squash all commits anyway that's fine

**Review by tdelabro — APPROVED** [review]

**Review by tdelabro — APPROVED** [review]

**Comment by tdelabro** [comment]
@peyha more conflicts sorry

**Comment by peyha** [comment]
Now rpc tests are not passing anymore

**Comment by tdelabro** [comment]
> Now rpc tests are not passing anymore

It's already the case in main, not you bad

**Comment by peyha** [comment]
> It's already the case in main, not you bad

Saw that every PR was failing shortly after sending my message

Hopefully first and not last PR merged on the repo :rocket: ready to take any related (or not) issue now that I am more familiar with the project
(and #1394 can be closed)

**Pull request merged**